File: e2e_tests/python/server_clients/lambda_function_url.py
"""
Lambda Function URL client for MCP servers running behind Lambda function URLs.

This client uses AWS SigV4 authentication to communicate with MCP servers
deployed as Lambda functions with function URLs enabled.
"""

import logging
from datetime import timedelta
from typing import Optional

import boto3
from botocore.exceptions import ClientError

from mcp.client.session import ClientSession
from mcp_lambda.client.streamable_http_sigv4 import streamablehttp_client_with_sigv4

from server_clients.server import Server


class LambdaFunctionUrlConfig:
    """Configuration for LambdaFunctionUrlClient."""

    def __init__(
        self,
        function_url: Optional[str] = None,
        stack_name: Optional[str] = None,
        stack_url_output_key: str = "FunctionUrl",
        region: str = "us-east-2",
        **kwargs,
    ):
        # Handle camelCase parameter names from JSON config
        self.function_url = kwargs.get("functionUrl", function_url)
        self.stack_name = kwargs.get("stackName", stack_name)
        self.stack_url_output_key = kwargs.get(
            "stackUrlOutputKey", stack_url_output_key
        )
        self.region = kwargs.get("region", region)


class LambdaFunctionUrlClient(Server):
    """
    Manages MCP server connections and tool execution for servers running behind
    Lambda function URLs with AWS SigV4 authentication.

    This client can lookup the function URL from a CloudFormation stack output
    instead of requiring the user to statically configure the URL.
    """

    def __init__(self, name: str, config: LambdaFunctionUrlConfig):
        # Convert config to dict for base class
        config_dict = {
            "function_url": config.function_url,
            "stack_name": config.stack_name,
            "stack_url_output_key": config.stack_url_output_key,
            "region": config.region,
        }
        super().__init__(name, config_dict)

        if not config.function_url and not config.stack_name:
            raise ValueError(
                "Either function_url must be provided or stack_name must be provided for CloudFormation lookup"
            )

        if config.function_url and config.stack_name:
            raise ValueError("Only one of function_url or stack_name can be provided")

        self.lambda_config = config
        self._transport_context = None
        self._session_context = None
        self._streams = None

    async def initialize(self) -> None:
        """Initialize the server connection with AWS SigV4 authentication."""
        try:
            # Determine the function URL
            function_url = self.lambda_config.function_url
            if self.lambda_config.stack_name:
                logging.debug("Retrieving function URL from CloudFormation...")
                function_url = await self._get_function_url_from_cloudformation()
                # Update the config with the resolved URL
                self.config["function_url"] = function_url
                self.lambda_config.function_url = function_url

            if not function_url:
                raise ValueError(
                    "The function_url must be a valid string and cannot be undefined."
                )

            logging.debug(f"Connecting to Lambda function URL: {function_url}")

            session = boto3.Session()
            credentials = session.get_credentials()
            if not credentials:
                raise ValueError(
                    "AWS credentials not found. Please configure your AWS credentials."
                )

            logging.debug("Creating transport with SigV4 authentication...")
            self._transport_context = streamablehttp_client_with_sigv4(
                url=function_url,
                credentials=credentials,
                service="lambda",
                region=self.lambda_config.region,
                timeout=timedelta(seconds=60),
            )

            # Enter the transport context
            self._streams = await self._transport_context.__aenter__()
            read_stream, write_stream, get_session_id = self._streams

            logging.debug("Creating MCP session...")
            self._session_context = ClientSession(read_stream, write_stream)
            self.session = await self._session_context.__aenter__()

            logging.debug("Initializing MCP session...")
            await self.session.initialize()
            logging.debug("MCP session initialized successfully")

        except Exception as error:
            logging.error(
                f"Error initializing Lambda function URL client {self.name}: {error}"
            )
            raise error

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit with proper cleanup."""
        try:
            if self._session_context:
                await self._session_context.__aexit__(exc_type, exc_val, exc_tb)
                self._session_context = None
                self.session = None

            if self._transport_context:
                await self._transport_context.__aexit__(exc_type, exc_val, exc_tb)
                self._transport_context = None
                self._streams = None
        except Exception as e:
            logging.error(f"Error during cleanup: {e}")

        # Call parent cleanup
        await super().__aexit__(exc_type, exc_val, exc_tb)

    async def _get_function_url_from_cloudformation(self) -> str:
        """Retrieve the Lambda function URL from CloudFormation stack outputs."""
        try:
            logging.debug(
                f"Retrieving function URL from CloudFormation stack: {self.lambda_config.stack_name}"
            )

            # Create CloudFormation client
            session = boto3.Session()
            cf_client = session.client(
                "cloudformation", region_name=self.lambda_config.region
            )

            response = cf_client.describe_stacks(
                StackName=self.lambda_config.stack_name
            )

            if not response.get("Stacks"):
                raise ValueError(
                    f"CloudFormation stack '{self.lambda_config.stack_name}' not found"
                )

            stack = response["Stacks"][0]
            if not stack.get("Outputs"):
                raise ValueError(
                    f"No outputs found in CloudFormation stack '{self.lambda_config.stack_name}'"
                )

            function_url_output = next(
                (
                    output
                    for output in stack["Outputs"]
                    if output["OutputKey"] == self.lambda_config.stack_url_output_key
                ),
                None,
            )

            if not function_url_output or not function_url_output.get("OutputValue"):
                raise ValueError(
                    f"Function URL output not found in CloudFormation stack. Output key: {self.lambda_config.stack_url_output_key}"
                )

            function_url = function_url_output["OutputValue"]
            logging.debug(f"Retrieved function URL: {function_url}")
            return function_url

        except ClientError as error:
            error_code = error.response["Error"]["Code"]
            if error_code == "ValidationException":
                raise ValueError(
                    f"CloudFormation stack '{self.lambda_config.stack_name}' does not exist or is not accessible"
                )
            elif error_code in ["AccessDenied", "UnauthorizedOperation"]:
                raise ValueError(
                    f"Insufficient permissions to access CloudFormation stack '{self.lambda_config.stack_name}'. "
                    "Ensure your AWS credentials have cloudformation:DescribeStacks permission."
                )
            else:
                raise ValueError(
                    f"Could not retrieve function URL from CloudFormation stack {self.lambda_config.stack_name}: {error}"
                )

        except Exception as error:
            logging.error(
                f"Failed to retrieve function URL from CloudFormation:", error
            )
            raise ValueError(
                f"Could not retrieve function URL from CloudFormation stack {self.lambda_config.stack_name}: {error}"
            )

File: examples/chatbots/typescript/src/server_clients/lambda_function_url.ts
import {
  CloudFormationClient,
  DescribeStacksCommand,
} from "@aws-sdk/client-cloudformation";
import { StreamableHTTPClientWithSigV4Transport } from "@aws/run-mcp-servers-with-aws-lambda";
import { Server } from "./server.js";
import logger from "../logger.js";

/**
 * Configuration for LambdaFunctionUrlClient
 */
export interface LambdaFunctionUrlConfig {
  /**
   * Direct Lambda function URL (if provided, stackName should not be used)
   */
  functionUrl?: string;

  /**
   * CloudFormation stack name to lookup function URL from (if provided, functionUrl should not be used)
   */
  stackName?: string;

  /**
   * CloudFormation stack output key for the function URL (default: "FunctionUrl")
   */
  stackUrlOutputKey?: string;

  /**
   * AWS region (default: "us-west-2")
   */
  region?: string;
}

/**
 * Manages MCP server connections and tool execution for servers running behind
 * Lambda function URLs with AWS SigV4 authentication.
 *
 * This client can lookup the function URL from a CloudFormation stack output
 * instead of requiring the user to statically configure the URL.
 */
export class LambdaFunctionUrlClient extends Server {
  private lambdaConfig: Required<LambdaFunctionUrlConfig>;

  constructor(name: string, config: LambdaFunctionUrlConfig) {
    // Set defaults and convert to required config
    const fullConfig: Required<LambdaFunctionUrlConfig> = {
      functionUrl: config.functionUrl || "",
      stackName: config.stackName || "",
      stackUrlOutputKey: config.stackUrlOutputKey || "FunctionUrl",
      region: config.region || "us-west-2",
    };

    super(name, fullConfig);
    this.lambdaConfig = fullConfig;

    if (!fullConfig.functionUrl && !fullConfig.stackName) {
      throw new Error(
        "Either functionUrl must be provided or stackName must be provided for CloudFormation lookup"
      );
    }

    if (fullConfig.functionUrl && fullConfig.stackName) {
      throw new Error("Only one of functionUrl or stackName can be provided");
    }
  }

  /**
   * Initialize the server connection with AWS SigV4 authentication.
   * @throws Error if initialization parameters are invalid
   * @throws Error if server fails to initialize
   */
  async initialize(): Promise<void> {
    try {
      // Determine the function URL
      let functionUrl = this.lambdaConfig.functionUrl;
      if (this.lambdaConfig.stackName) {
        logger.debug("Retrieving function URL from CloudFormation...");
        functionUrl = await this._getFunctionUrlFromCloudFormation();
        // Update the config with the resolved URL
        this.lambdaConfig.functionUrl = functionUrl;
      }

      if (!functionUrl) {
        throw new Error(
          "The functionUrl must be a valid string and cannot be undefined."
        );
      }

      logger.debug(`Connecting to Lambda function URL: ${functionUrl}`);

      const transport = new StreamableHTTPClientWithSigV4Transport(
        new URL(functionUrl),
        {
          service: "lambda",
          region: this.lambdaConfig.region,
        }
      );

      await this.client.connect(transport);
      logger.debug("MCP session initialized successfully");
    } catch (error) {
      logger.error(
        `Error initializing Lambda function URL client ${this.name}: ${error}`
      );
      throw error;
    }
  }

  /**
   * Retrieve the Lambda function URL from CloudFormation stack outputs.
   */
  private async _getFunctionUrlFromCloudFormation(): Promise<string> {
    try {
      logger.debug(
        `Retrieving function URL from CloudFormation stack: ${this.lambdaConfig.stackName}`
      );

      const cfClient = new CloudFormationClient({
        region: this.lambdaConfig.region,
      });

      const command = new DescribeStacksCommand({
        StackName: this.lambdaConfig.stackName,
      });

      const response = await cfClient.send(command);

      if (!response.Stacks || response.Stacks.length === 0) {
        throw new Error(
          `CloudFormation stack '${this.lambdaConfig.stackName}' not found`
        );
      }

      const stack = response.Stacks[0];
      if (!stack.Outputs) {
        throw new Error(
          `No outputs found in CloudFormation stack '${this.lambdaConfig.stackName}'`
        );
      }

      const functionUrlOutput = stack.Outputs.find(
        (output) => output.OutputKey === this.lambdaConfig.stackUrlOutputKey
      );

      if (!functionUrlOutput || !functionUrlOutput.OutputValue) {
        throw new Error(
          `Function URL output not found in CloudFormation stack. Output key: ${this.lambdaConfig.stackUrlOutputKey}`
        );
      }

      const functionUrl = functionUrlOutput.OutputValue;
      logger.debug(`Retrieved function URL: ${functionUrl}`);
      return functionUrl;
    } catch (error: any) {
      if (error.name === "ValidationException") {
        throw new Error(
          `CloudFormation stack '${this.lambdaConfig.stackName}' does not exist or is not accessible`
        );
      } else if (
        error.name === "AccessDenied" ||
        error.name === "UnauthorizedOperation"
      ) {
        throw new Error(
          `Insufficient permissions to access CloudFormation stack '${this.lambdaConfig.stackName}'. ` +
            "Ensure your AWS credentials have cloudformation:DescribeStacks permission."
        );
      } else {
        logger.error(
          "Failed to retrieve function URL from CloudFormation:",
          error
        );
        throw new Error(
          `Could not retrieve function URL from CloudFormation stack ${this.lambdaConfig.stackName}: ${error.message}`
        );
      }
    }
  }
}
